---
title: "Gviz for PDE4DIP"
author: "C.J. Liu"
date: "1/17/2017"
output:
  html_document:
    depth: 3
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: no
    word_document:
      toc: yes
  word_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```
# Gviz for PDE4DIP
```{r load library}
suppressMessages(require(Gviz))
suppressMessages(require(dplyr))
suppressMessages(require(GenomicRanges))
suppressMessages(require(biomaRt))
suppressMessages(require(tidyr))
```

## Level-1 Ideogram
Set genome as hg38
Chomosome as chromosome 1
```{r leve-1}
gen = 'hg38'
chr = 'chr1'
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
plotTracks(itrack, from = 148651032, to = 148671032)
```

## Level-2 Coordinates
```{r level-2}
gtrack <- GenomeAxisTrack()
plotTracks(list(itrack, gtrack),from = 148651032, to = 148671032)
```

## Level-4 Gene Model of PDE4DIP
The point mutation in the NBPF14 intron. also plot NBPF14 gene model.

### Get NBPF14 and PDE4DIP gene medels
```{r biomart}
GENES = useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = "useast.ensembl.org")
GENES.ATTRIBUTES <- listAttributes(GENES)
GENES.FILTERS <- listFilters(GENES)
```
PDEDIP ensembl ID is **ENSG00000178104**  
NBPF14 ensembl ID is **ENSG00000270629**
```{r Gene Model}
#pde4dip dbxref
PDE4DIP.dbxref <- getBM(attributes = c('hgnc_symbol', 'ensembl_transcript_id','refseq_mrna', 'ucsc'), filters = c('ensembl_gene_id','transcript_biotype'), values = list(ensembl_gene_id = 'ENSG00000178104',transcript_biotype = 'protein_coding'), mart = GENES)

# remove duplicates of ensembl_transcript_id create by multiple mapping to Refseq ID
PDE4DIP.dbxref <- distinct(PDE4DIP.dbxref, ensembl_transcript_id, .keep_all = T)

# get pde4dip all exons regions
PDE4DIP.exons <- getBM(attributes = c('chromosome_name', 'exon_chrom_start', 'exon_chrom_end','strand','ensembl_gene_id','ensembl_transcript_id','transcript_biotype','ensembl_exon_id'), filters = c('ensembl_gene_id','transcript_biotype'), values = list(ensembl_gene_id = 'ENSG00000178104',transcript_biotype = 'protein_coding'), mart = GENES)
PDE4DIP.exons <- mutate(PDE4DIP.exons, width = exon_chrom_end - exon_chrom_start + 1)

# merge annotations to exons regions
PDE4DIP.geneModel <- left_join(PDE4DIP.exons, PDE4DIP.dbxref, by = 'ensembl_transcript_id')

# reorder PDE4DIP gene model names to satisfy GeneRegionTrack construction.
PDE4DIP.geneModel <- PDE4DIP.geneModel %>% transmute(chromosome = paste('chr', chromosome_name, sep = ''), start = exon_chrom_start, end = exon_chrom_end, width = width, strand = gsub('1','+', strand), feature = transcript_biotype, gene = ensembl_gene_id, exon  = ensembl_exon_id, transcript = ensembl_transcript_id, symbol = hgnc_symbol, refseq = refseq_mrna, ucsc = ucsc)

# Construct pde4dip gene region track
PDE4DIP.grtrack <- GeneRegionTrack(PDE4DIP.geneModel, genome = gen, chromosome = chr, name = 'PDE4DIP', transcriptAnnotation = "transcript", background.title = "brown")
plotTracks(list(itrack, gtrack, PDE4DIP.grtrack))




#NBPF14 dbxref
NBPF14.dbxref <- getBM(attributes = c('hgnc_symbol', 'ensembl_transcript_id','refseq_mrna', 'ucsc'), filters = c('ensembl_gene_id'), values = list(ensembl_gene_id = 'ENSG00000270629'), mart = GENES)

# remove duplicates of ensembl_transcript_id create by multiple mapping to Refseq ID
NBPF14.dbxref <- distinct(NBPF14.dbxref, ensembl_transcript_id, .keep_all = T)

# get NBPF14 all exons regions
NBPF14.exons <- getBM(attributes = c('chromosome_name', 'exon_chrom_start', 'exon_chrom_end','strand','ensembl_gene_id','ensembl_transcript_id','transcript_biotype','ensembl_exon_id'), filters = c('ensembl_gene_id'), values = list(ensembl_gene_id = 'ENSG00000270629'), mart = GENES)
NBPF14.exons <- mutate(NBPF14.exons, width = exon_chrom_end - exon_chrom_start + 1)

# merge annotations to exons regions
NBPF14.geneModel <- left_join(NBPF14.exons, NBPF14.dbxref, by = 'ensembl_transcript_id')

# reorder NBPF14 gene model names to satisfy GeneRegionTrack construction.
NBPF14.geneModel <- NBPF14.geneModel %>% transmute(chromosome = paste('chr', chromosome_name, sep = ''), start = exon_chrom_start, end = exon_chrom_end, width = width, strand = gsub('-1','-', strand), feature = transcript_biotype, gene = ensembl_gene_id, exon  = ensembl_exon_id, transcript = ensembl_transcript_id, symbol = hgnc_symbol, refseq = refseq_mrna, ucsc = ucsc)

# Construct NBPF14 gene region track
NBPF14.grtrack <- GeneRegionTrack(NBPF14.geneModel, genome = gen, chromosome = chr, name = 'NBPF14', transcriptAnnotation = "transcript", background.title = "brown")
plotTracks(list(itrack, gtrack, NBPF14.grtrack))

# combine PDE4DIP and NBPF14
geneModels <- rbind(PDE4DIP.geneModel, NBPF14.geneModel)
geneModels <- geneModels %>% filter(transcript %in% c('ENST00000593495','ENST00000606877', 'ENST00000479408','ENST00000313431','ENST00000529945','ENST00000491426'))
grTrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, name = 'PDE4DIP', transcriptAnnotation = "transcript", background.title = "brown")

```

### Plot PDE4DIP and NBPF14 transcripts
```{r level-4}
# mutation is in the ENST00000593495 1/4 intron
# last exons is ENSE00003126940 from 148607038 to 148607159
# PDE4DIP last exon position is max(PDE4DIP.geneModel$end)
plotTracks(list(itrack, gtrack, grTrack), from = 148517038, to = 149033019)
```

## Level-3 TF binding/regulation position

1. Extract all transcripts of PDE4DIP
2. Find TF of 1000 up stream transcripts.
3. Mutation of ENSR bind motif for USF1, and TCF7L2

### Get PDE4DIP all transcripts and 1000 upstream positions.

### ENSR00000013230 binding TFs
chr1    148660601       148664799       ENSR00000013230

chr1    148660601       148660614       USF1
chr1    148661025       148661133       TCF7L2

chr1:148660428-148660614 USF1[186]
AAGAAGATTCCTTAATTCTGTCAATAGAATAATGATCCAGTTATATACCAAAGAGCAGATAGAAGACCTGGCTATGGAGTCATGTGACTCTGCATTAGATCTCATGGGGTCTCTGTCTCAGTCTCCACAAATAAAAACATGGCCTTGTCTCTATAAAAGATACAATGTACATAGAAGTGTTTTGCA

chr1:148661025-148661133 TCF7L2[108]
TGACTT[**A>T**]TCTTTTGCTATGGGTACACTGTCTTTTCATGTGGCCTCAAAGAAATAGTTGAGCCATTATTGTTGTTGATAATGACGAGCAGATGCTGATGGGATGATTGC

```{r transcripts and upstream}
PDE4DIP.transcripts <- getBM(attributes = c('chromosome_name', 'strand','transcript_start','transcript_end','ensembl_gene_id','ensembl_transcript_id','refseq_mrna','ucsc'), filters = c('ensembl_gene_id','transcript_biotype'), values = list(ensembl_gene_id = 'ENSG00000178104',transcript_biotype = 'protein_coding'), mart = GENES)

PDE4DIP.transcripts <- PDE4DIP.transcripts %>% transmute(chromosome = paste('chr', chromosome_name, sep = ''), strand = gsub('1','+', strand), upstream = transcript_start - 1000, start = transcript_start, end = transcript_end, gene = ensembl_gene_id, transcript=ensembl_transcript_id, refseq = refseq_mrna, ucsc = ucsc)

PDE4DIP.transcripts <- PDE4DIP.transcripts %>% distinct(transcript, .keep_all = T)

PDE4DIP.transcripts.separate <- PDE4DIP.transcripts %>% separate(ucsc, into=c("ucsc", 'version'))

# write PDE4DIP.transcripts.separate to /home/cliu18/liucj/reference/TFs
write.table(PDE4DIP.transcripts.separate, file = '/home/cliu18/liucj/reference/TFs/PDE4DIP.transcripts.separate', row.names = F, quote = F, sep = "\t")

# bedtools intersect -a hTFtarget_datasetfilteredpeak.bed.sortByChrom -b PDE4DIP.transcripts.separate.bed > PDE4DIP.transcripts.separate.bed.motif

# laod transcript motif
PDE4DIP.transcripts.separate.motif <- read.table(file = '/home/cliu18/liucj/reference/TFs/PDE4DIP.transcripts.separate.bed.motif.uniq', sep = "\t")

colnames(PDE4DIP.transcripts.separate.motif) <- c("chromosome", "start", "end", "tf")

PDE4DIP.transcripts.separate.motif <- PDE4DIP.transcripts.separate.motif %>% mutate(strand = "*")

#chr1    148660601       148660614       USF1
#chr1    148661025       148661133       TCF7L2
#chr1    148660601       148664799       ENSR00000013230
PDE4DIP.ensr <- data.frame(chromosome = c('chr1', 'chr1', 'chr1'), start = c(148660601, 148661025, 148660601), end = c(148660614,148661133,148664799), tf = c('USF1', 'TCF7L2','ENSR00000013230'), strand = c("*","*","*"))

PDE4DIP.transcripts.separate.motif <- rbind(PDE4DIP.ensr, PDE4DIP.transcripts.separate.motif)

# there are too many binding on the same position, draw only part of them.
PDE4DIP.transcripts.separate.motif <- PDE4DIP.transcripts.separate.motif %>% filter(tf %in% c("USF1", "TCF7L2", "EP300","BRCA1","CDK9","MYC","ENSR00000013230"))
# PDE4DIP.transcripts.separate.motif <- PDE4DIP.transcripts.separate.motif %>% filter(!tf  %in% c("SPI1","CTCF","POLR2A","YY1","TFAP2C"))

PDE4DIP.transcripts.separate.motif.grange <- makeGRangesFromDataFrame(PDE4DIP.transcripts.separate.motif,keep.extra.columns=TRUE)

atrack <- AnnotationTrack(PDE4DIP.transcripts.separate.motif.grange, name = "TF Binding",cex.group = 0.6, background.title = "blue")
feature(atrack) <- as.character(PDE4DIP.transcripts.separate.motif.grange$tf)
```


```{r level-3}
plotTracks(list(itrack, gtrack,atrack, grTrack), from = 148517038, to = 149033019, groupAnnotation = "feature")
```


## Level-5 Load expression data
Expression data was generated by scripts `nearestGeneExpression2`.
```{r transcripts expression}
# load PDE4DIP transcripts expression
PDE4DIP.transcripts.expression <- read.table('/extraspace/TCGA/WXS_RAW/BRCA/Firehose/BRCA/mRNASeq/gdac.broadinstitute.org_BRCA.Merge_rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_isoforms_normalized__data.Level_3.2016012800.0.0/pde4dip.all.transcript.expression', header = T, sep = "\t")

PDE4DIP.transcripts.expression <- separate(PDE4DIP.transcripts.expression, Hybridization.REF, into = c('ucsc','version'))

PDE4DIP.transcripts.expression <- PDE4DIP.transcripts.expression %>% gather(barcode, rsem, -c(ucsc, version))

PDE4DIP.transcripts.expression$barcode <- substring(PDE4DIP.transcripts.expression$barcode,1,16)

# gene expression
PDE4DIP.gene.expression <- read.table(file = file.path('/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression', 'PDE4DIP.expression.txt'), header = T, sep = "\t" )

PDE4DIP.gene.expression <- PDE4DIP.gene.expression %>% distinct(barcode, .keep_all = T)

PDE4DIP.transcripts.expression <-  inner_join(PDE4DIP.gene.expression, PDE4DIP.transcripts.expression, by = 'barcode')

PDE4DIP.transcripts.expression <- inner_join(PDE4DIP.transcripts.expression, PDE4DIP.transcripts.separate, by = 'ucsc')


PDE4DIP.transcripts.expression.one <- PDE4DIP.transcripts.expression %>% filter(transcript %in% c('ENST00000593495','ENST00000606877', 'ENST00000479408','ENST00000313431','ENST00000529945','ENST00000491426'))

PDE4DIP.transcripts.expression.one.dtrack <- PDE4DIP.transcripts.expression.one %>% dplyr::select(chromosome, start = start, end, strand,barcode, rsem)

PDE4DIP.transcripts.expression.one.dtrack <- PDE4DIP.transcripts.expression.one.dtrack %>% mutate(start = start - 30000) %>% spread(barcode, rsem)

PDE4DIP.transcripts.expression.one.dtrack <- makeGRangesFromDataFrame(PDE4DIP.transcripts.expression.one.dtrack, keep.extra.columns=TRUE)

dTrack <- DataTrack(PDE4DIP.transcripts.expression.one.dtrack, name = 'Isoform Expression', groups = PDE4DIP.transcripts.expression.one$type, type = 'boxplot', legend = T)
```


```{r level-5}

plotTracks(list(itrack, gtrack,atrack, grTrack,dTrack), from = 148517038, to = 149033019, groupAnnotation = "feature")
plotTracks(list(itrack, gtrack,atrack, grTrack,dTrack), from = 148900000, to = 149033019, groupAnnotation = "feature")
plotTracks(list(itrack, gtrack,atrack, grTrack,dTrack), from = 148517038, to = 148700000, groupAnnotation = "feature")

```

## Level-6 Hightlight mutation
The point mutation **chr1:148661032A>T**
```{r level-6}

ht <- HighlightTrack(trackList = list(atrack,grTrack,dTrack), start = 148661032, width = 1, chromosome = 1 )

# The whole picture
plotTracks(list(itrack, gtrack, ht), from = 148517038, to = 149033019, groupAnnotation = "feature")
pdf(file = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/mutationTFBindingTranscriptsExpression.whole.pdf")
plotTracks(list(itrack, gtrack, ht), from = 148517038, to = 149033019, groupAnnotation = "feature")
dev.off()

# The mutation picture
plotTracks(list(itrack, gtrack,ht), from = 148657038, to = 148670000, groupAnnotation = "feature")
pdf(file = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/mutationTFBindingTranscriptsExpression.mutation.pdf")
plotTracks(list(itrack, gtrack,ht), from = 148657038, to = 148670000, groupAnnotation = "feature")
dev.off()

# The PDE4DIP transcripts
plotTracks(list(itrack, gtrack,atrack, grTrack,dTrack), from = 148940000, to = 148963019, groupAnnotation = "feature")
pdf(file = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/mutationTFBindingTranscriptsExpression.transcripts.pdf")
plotTracks(list(itrack, gtrack,atrack, grTrack,dTrack), from = 148940000, to = 148963019, groupAnnotation = "feature")
dev.off()


```
















