---
title: "nearestGeneExpression"
author: "C.J. Liu"
date: "12/23/2016"
output:
  html_document:
    depth: 3
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: no
    word_document:
      toc: yes
  word_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r load library}
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
library(knitr)
```

## Load nearestProteinCodingGene
```{r load data}
# load nearest gene
nearestProteinCodingGene <- read.table(file="/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/6.mapMutationToGenes/SNP/nearestProteinCodingGene.txt", header = T, sep = "\t")

coln <- c("uuid", "chrom", "pos", "depth", "type", "barcode")

#tmp.txt <- read.table(file.path("/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/5.recalibrateRecurrencyWithCoverage/SNP/refined_goc", paste(paste(recurr, paste("chr", chrom, sep = ""), pos, sep = "_"), ".10.mtwt.barcode", sep = "")), sep = "\t", header = F)

loadBarcode <- function(x, directory){
  path <- file.path(directory,paste(paste(as.numeric(x["recurr"]), paste("chr", x["chrom"], sep = ""), as.numeric(x["pos"]), sep = "_"), ".10.mtwt.barcode", sep = ""))
  if(file.exists(path)){
    #print(path)
    tmp <- read.table(file = path, header = F, sep = "\t")
    colnames(tmp) <- coln
    tmp$barcode <- gsub("-", ".", tmp$barcode)
    tmp$ensembl_gene_id <- x["ensembl_gene_id"]
    tmp$hgnc_symbol <- x["hgnc_symbol"]
    return(tmp)
  }
}
nearestProteinCodingGene.distinct <- distinct(nearestProteinCodingGene, ensembl_gene_id, .keep_all = T)
expr.candidate <- apply(nearestProteinCodingGene.distinct, 1, loadBarcode, directory = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/5.recalibrateRecurrencyWithCoverage/SNP/refined_goc")
names(expr.candidate) <- nearestProteinCodingGene.distinct$ensembl_gene_id

# load expression data
expr <- read.table(file = "/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/brca.fpkm.expr.cases.genes.txt", header = T, sep = "\t")

getGeneExpression <- function(x){
  tmp <- dplyr::select(expr, one_of(c("ensid",x$barcode))) %>% filter(ensid %in% x$ensembl_gene_id)
  tmp <-as.data.frame(t(tmp))
  colnames(tmp) <- "expression"
  tmp$expression <- as.numeric(tmp$expression)
  tmp$barcode <- rownames(tmp)
  left_join(x,tmp, by = "barcode")
}
# test <- head(expr.candidate,1)
# lapply(test, getGeneExpression)
total.expr <- lapply(expr.candidate,getGeneExpression)

#pdf(file.path("/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression","all.gene_expression.pdf"), width = 20, height = 20) 

for(gene in total.expr){
  p.value <- wilcox.test(data=gene, expression ~ type)$p.value
  if(is.na(p.value)) next
  if(p.value < 0.05){
    p <- ggplot(data = gene, aes(x = type, y = expression)) + geom_boxplot(aes(color = type)) + geom_jitter(aes(color= type), position = position_jitter(0.1)) + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(gene$hgnc_symbol[1], "MT vs. WT p-value", round(p.value,3),"(wilcox test)"), subtitle = "")
    print(p)
    ggsave(paste(gene$hgnc_symbol[1], "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")
    
   }
}

PDE4DIP <- total.expr$ENSG00000178104[, c(5,6,7,8,9)]
PDE4DIP.normal <- data.frame(type = c("NM", "NM"), barcode = c('1','1'), ensembl_gene_id = rep("ENSG00000178104", 2), hgnc_symbol = rep("PDE4DIP",2), expression = c(5.67827554477, 2.51608176292))
PDE4DIP.expression <- rbind(PDE4DIP,PDE4DIP.normal)


p <- ggplot(data = PDE4DIP.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type)) + geom_jitter(aes(color= type), position = position_jitter(0.1)) + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(PDE4DIP.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = "")
print(p)
ggsave(paste("PDE4DIP","normal", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")

# all normal sample expression
PDE4DIP.normal <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/PDE4DIP.normalName.expr.txt", header = T)
PDE4DIP.normal <- data.frame(type = "NM", barcode = colnames(PDE4DIP.normal), ensembl_gene_id = "ENSG00000178104",hgnc_symbol = "PDE4DIP", expression = as.numeric(PDE4DIP.normal[1,]))
PDE4DIP.expression <- rbind(PDE4DIP,PDE4DIP.normal)
PDE4DIP.expression <- filter(PDE4DIP.expression, barcode != 'ensid')

PDE4DIP.MT.WT <- filter(PDE4DIP.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = PDE4DIP.MT.WT, expression ~ type)$p.value
PDE4DIP.WT.NM <- filter(PDE4DIP.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = PDE4DIP.WT.NM, expression ~ type)$p.value
p <- ggplot(data = PDE4DIP.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(PDE4DIP.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))

print(p)
ggsave(paste("PDE4DIP","normal.all", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")
# save PDE4DIP.expression to file PDE4DIP.expression.txt
write.table(PDE4DIP.expression, file = file.path("/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/","PDE4DIP.expression.txt"), row.names = F, quote = F, col.names = T, sep = "\t")


# USF1 expression
USF1.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/USF1.all.expr.txt", header = T, sep = "\t")
USF1.expression <- mutate(PDE4DIP.expression, expression = as.numeric(USF1.expr[,PDE4DIP.expression$barcode]), ensembl_gene_id= "ENSG00000158773", hgnc_symbol = "USF1")

USF1.MT.WT <- filter(USF1.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = USF1.MT.WT, expression ~ type)$p.value
USF1.WT.NM <- filter(USF1.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = USF1.WT.NM, expression ~ type)$p.value

p <- ggplot(data = USF1.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(USF1.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))

print(p)
ggsave(paste("USF1", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")

# TCF7L2 expression
TCF7L2.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/TCF7L2.all.expr.txt", header = T, sep = "\t")
TCF7L2.expression <- mutate(PDE4DIP.expression, expression = as.numeric(TCF7L2.expr[,PDE4DIP.expression$barcode]), ensembl_gene_id= "ENSG00000148737", hgnc_symbol = "TCF7L2")

TCF7L2.MT.WT <- filter(TCF7L2.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = TCF7L2.MT.WT, expression ~ type)$p.value
TCF7L2.WT.NM <- filter(TCF7L2.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = TCF7L2.WT.NM, expression ~ type)$p.value

p <- ggplot(data = TCF7L2.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(TCF7L2.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))


print(p)
ggsave(paste("TCF7L2", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")

# MYC expression
MYC.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/MYC.all.expr.txt", header = T, sep = "\t")
MYC.expression <- mutate(PDE4DIP.expression, expression = as.numeric(MYC.expr[,PDE4DIP.expression$barcode]), ensembl_gene_id= "ENSG00000136997", hgnc_symbol = "MYC")

MYC.MT.WT <- filter(MYC.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = MYC.MT.WT, expression ~ type)$p.value
MYC.WT.NM <- filter(MYC.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = MYC.WT.NM, expression ~ type)$p.value

p <- ggplot(data = MYC.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(MYC.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))


print(p)
ggsave(paste("MYC", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")


# CTNNB1 expression
CTNNB1.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/CTNNB1.all.expr.txt", header = T, sep = "\t")
CTNNB1.expression <- mutate(PDE4DIP.expression, expression = as.numeric(CTNNB1.expr[,PDE4DIP.expression$barcode]), ensembl_gene_id= "ENSG00000168036", hgnc_symbol = "CTNNB1")

CTNNB1.MT.WT <- filter(CTNNB1.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = CTNNB1.MT.WT, expression ~ type)$p.value
CTNNB1.WT.NM <- filter(CTNNB1.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = CTNNB1.WT.NM, expression ~ type)$p.value

p <- ggplot(data = CTNNB1.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(CTNNB1.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))


print(p)
ggsave(paste("CTNNB1", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")

# NOTCH2 expression
NOTCH2.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/NOTCH2.all.expr.txt", header = T, sep = "\t")
NOTCH2.expression <- mutate(PDE4DIP.expression, expression = as.numeric(NOTCH2.expr[,PDE4DIP.expression$barcode]), ensembl_gene_id= "ENSG00000134250", hgnc_symbol = "NOTCH2")

NOTCH2.MT.WT <- filter(NOTCH2.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = NOTCH2.MT.WT, expression ~ type)$p.value
NOTCH2.WT.NM <- filter(NOTCH2.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = NOTCH2.WT.NM, expression ~ type)$p.value

p <- ggplot(data = NOTCH2.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(NOTCH2.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))


print(p)
ggsave(paste("NOTCH2", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")

# TLE1 expression
TLE1.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/TLE1.all.expr.txt", header = T, sep = "\t")
TLE1.expression <- mutate(PDE4DIP.expression, expression = as.numeric(TLE1.expr[,PDE4DIP.expression$barcode]), ensembl_gene_id= "ENSG00000196781", hgnc_symbol = "TLE1")

TLE1.MT.WT <- filter(TLE1.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = TLE1.MT.WT, expression ~ type)$p.value
TLE1.WT.NM <- filter(TLE1.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = TLE1.WT.NM, expression ~ type)$p.value

p <- ggplot(data = TLE1.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(TLE1.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))


print(p)
ggsave(paste("TLE1", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")

# EP300 expression
EP300.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/EP300.all.expr.txt", header = T, sep = "\t")
EP300.expression <- mutate(PDE4DIP.expression, expression = as.numeric(EP300.expr[,PDE4DIP.expression$barcode]), ensembl_gene_id= "ENSG00000100393", hgnc_symbol = "EP300")

EP300.MT.WT <- filter(EP300.expression, type %in% c("MT","WT"))
mtwt.wilcox.pvalue <- wilcox.test(data = EP300.MT.WT, expression ~ type)$p.value
EP300.WT.NM <- filter(EP300.expression, type %in% c("WT","NM"))
wtnm.wilcox.pvalue <- wilcox.test(data = EP300.WT.NM, expression ~ type)$p.value

p <- ggplot(data = EP300.expression, aes(x = type, y = expression)) + geom_boxplot(aes(color = type))  + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(EP300.expression$hgnc_symbol[1], "expression in Tumor MT,WT and Normal samples"), subtitle = paste("MT vs. WT p-value",round(mtwt.wilcox.pvalue,3),"and WT vs. NM p-value", round(wtnm.wilcox.pvalue,3)))


print(p)
ggsave(paste("EP300", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")



# CTCF expresison
CTCF.expr <- read.table("/extraspace/TCGA/WXS_RAW/BRCA/RNA-seq/CTCF.all.expr.txt", header = T, sep = "\t")
CTCF <- mutate(PDE4DIP, expression = as.numeric(CTCF.expr[,PDE4DIP$barcode]), ensembl_gene_id= "ENSG00000158773", hgnc_symbol = "CTCF")

p.value = wilcox.test(data=CTCF, expression ~ type)$p.value

p <- ggplot(data = CTCF, aes(x = type, y = expression)) + geom_boxplot(aes(color = type)) + geom_jitter(aes(color= type), position = position_jitter(0.1)) + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(CTCF$hgnc_symbol, "MT vs. WT p-value", round(p.value,3),"(wilcox test)"), subtitle = "")
print(p)
ggsave(paste("CTCF", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")



#p.value = wilcox.test(data=NBPF14, expression ~ type)$p.value
#p <- ggplot(data = NBPF14, aes(x = type, y = expression)) + geom_boxplot(aes(color = type)) + geom_jitter(aes(color= type), position = position_jitter(0.1)) + scale_color_brewer(palette="Dark2") + theme_minimal() + labs(x = "Type", y = "Expression (RPKM)", title = paste(NBPF14$hgnc_symbol, "MT vs. WT p-value", round(p.value,3),"(wilcox test)"), subtitle = "")


#ggsave(paste("NBPF14", "png", sep = "."), plot = p, device = "png", path = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/", width = 15, height = 15, units = "cm")

save.image(file = "/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/8.expression/expressionCached.RData")
```






